revocation/internal/crl/crl.go

```diff
@@ -188,38 +279,73 @@ func validate(crl *x509.RevocationList, issuer *x509.Certificate) error {
 }
 
 // checkRevocation checks if the certificate is revoked or not
-func checkRevocation(cert *x509.Certificate, baseCRL *x509.RevocationList, signingTime time.Time, crlURL string) (*result.ServerResult, error) {
+func checkRevocation(cert *x509.Certificate, b *crl.Bundle, signingTime time.Time, crlURL string) (*result.ServerResult, error) {
 	if cert == nil {
 		return nil, errors.New("certificate cannot be nil")
 	}
-
-	if baseCRL == nil {
+	if b == nil {
+		return nil, errors.New("CRL bundle cannot be nil")
+	}
+	if b.BaseCRL == nil {
 		return nil, errors.New("baseCRL cannot be nil")
 	}
 
-	for _, revocationEntry := range baseCRL.RevokedCertificateEntries {
-		if revocationEntry.SerialNumber.Cmp(cert.SerialNumber) == 0 {
-			extensions, err := parseEntryExtensions(revocationEntry)
-			if err != nil {
-				return nil, err
-			}
+	revocationListBundle := []*[]x509.RevocationListEntry{&b.BaseCRL.RevokedCertificateEntries}
+	if b.DeltaCRL != nil {
+		revocationListBundle = append(revocationListBundle, &b.DeltaCRL.RevokedCertificateEntries)
+	}
 
-			// validate signingTime and invalidityDate
-			if !signingTime.IsZero() && !extensions.invalidityDate.IsZero() &&
-				signingTime.Before(extensions.invalidityDate) {
-				// signing time is before the invalidity date which means the
-				// certificate is not revoked at the time of signing.
-				break
+	// latestTempRevokedEntry contains the most recent revocation entry with
+	// reasons such as CertificateHold or RemoveFromCRL.
+	//
+	// If the certificate is revoked with CertificateHold, it is temporarily
+	// revoked. If the certificate is shown in the CRL with RemoveFromCRL,
+	// it is unrevoked.
+	var latestTempRevokedEntry *x509.RevocationListEntry
+
+	// iterate over all the entries in the base and delta CRLs
+	for _, revocationList := range revocationListBundle {
+		for i, revocationEntry := range *revocationList {
+			if revocationEntry.SerialNumber.Cmp(cert.SerialNumber) == 0 {
+				extensions, err := parseEntryExtensions(revocationEntry)
+				if err != nil {
+					return nil, err
+				}
+
+				// validate signingTime and invalidityDate
+				if !signingTime.IsZero() && !extensions.invalidityDate.IsZero() &&
+					signingTime.Before(extensions.invalidityDate) {
+					// signing time is before the invalidity date which means the
+					// certificate is not revoked at the time of signing.
+					break
+				}
+
+				switch revocationEntry.ReasonCode {
+				case int(reasonCodeCertificateHold), int(reasonCodeRemoveFromCRL):
```

Comment:
nit: instead of converting `reasonCodeCertificateHold` to `int`, consider converting `revocationEntry.ReasonCode`